spring-boot-project/spring-boot-tools/spring-boot-configuration-processor/src/main/java/org/springframework/boot/configurationprocessor/fieldvalues/javac/JavaCompilerFieldValuesParser.java

```diff
@@ -165,7 +167,11 @@ private Object getValue(VariableTree variable) throws Exception {
 			Class<?> wrapperType = WRAPPER_TYPES.get(variable.getType());
 			Object defaultValue = DEFAULT_TYPE_VALUES.get(wrapperType);
 			if (initializer != null) {
-				return getValue(initializer, defaultValue);
+			    Object initializerResolvedValue = getValue(initializer, defaultValue);
```

Comment:
It needs to be a bit more precise than that. I think the idea is that we want to fail if we have a `FieldValuesParser` available and ignore if we don't.  We we could do is changing the contract to return a `ValueWrapper` and just return that here. Our fallback implementation when we couldn't find an implementation should then return null. 

Note that we probably want to differentiate between `null` as "detected as null" and `null` as "couldn't figure out". This could be an attribute on `ValueWrapper` or something.

Reply:
Thx for the direction @snicoll . It makes sense. I will get back to this in the next couple days.

Reply:
@snicoll 

> It needs to be a bit more precise than that. I think the idea is that we want to fail if we have a FieldValuesParser available and ignore if we don't. 

So, when no FVP is available (ie the env is not an `com.sun.tools.javac.processing.JavacProcessingEnvironment`) the [default FVP is NONE](https://github.com/spring-projects/spring-boot/blob/29637706a9978bd40b7a5593885ce279e168f17f/spring-boot-project/spring-boot-tools/spring-boot-configuration-processor/src/main/java/org/springframework/boot/configurationprocessor/MetadataGenerationEnvironment.java#L125) which simply returns an empty map. Because this code is in the FVP it will only apply when the FVP (itself) is available. 

> We we could do is changing the contract to return a ValueWrapper and just return that here. Our fallback implementation when we couldn't find an implementation should then return null.

Again, the fallback impl currently is an empty map. 

I may be missing something here - please clarify.

Reply:
> Again, the fallback impl currently is an empty map.

I am aware of that and I am asking you to change the contract. You can't know what fields are available if you don't have an implementation, but you could change the contract so that, if you're being requested a field with name "foo" you can then return `null` or a `ValueWrapper` that holds additional information as I've suggested. So instead of a map a POJO that can tell you if the result contains a field with a given name and another method that gives you a `ValueWrapper` based on a field name.

Reply:
I wanted to submit what I have so far before calling it a day as I know you 7hr ahead of me @snicoll . 

I have not moved from map to POJO and changed the contract yet as the map (w/ its ValueWrapper values) pretty much fulfills the above mentioned contract. I don't mind moving to POJO but am curious if you still think it will add value to the current proposal.

I am now marching through the JCFPV tests and updating it. I will continue that tomorrow.